### PR TITLE
Fixed #27542 -- Made Client.force_login() skip auth backends missing get_user().

### PR DESCRIPTION
**Pull request by desecho**
https://code.djangoproject.com/ticket/27542

### CODE REVIEW

File: django/test/client.py [hunk 1]
```diff
@@ -631,8 +631,14 @@ def login(self, **credentials):
             return False
 
     def force_login(self, user, backend=None):
+        def get_backend():
+            from django.contrib.auth import load_backend
+            for backend_path in settings.AUTHENTICATION_BACKENDS:
+                backend = load_backend(backend_path)
+                if hasattr(backend, 'get_user') and callable(backend.get_user):
+                    return backend_path
```
Comment (desecho) [inline django/test/client.py]: @orf, I've added Exception.

Comment (timgraham) [inline django/test/client.py]: I don't think there's a use case for `settings.AUTHENTICATION_BACKENDS` without any backends that have a `get_user()` method.

File: django/test/client.py [hunk 2]
```diff
@@ -631,8 +632,15 @@ def login(self, **credentials):
             return False
 
     def force_login(self, user, backend=None):
+        def get_backend():
+            from django.contrib.auth import load_backend
+            for backend_path in settings.AUTHENTICATION_BACKENDS:
+                backend = load_backend(backend_path)
+                if hasattr(backend, 'get_user') and callable(backend.get_user):
```
Comment (timgraham) [inline django/test/client.py]: Did you have a use case for `backend.get_user` not being callable? Do any tests fail if that's removed?

File: tests/auth_tests/test_auth_backends.py [hunk 3]
```diff
@@ -536,6 +536,10 @@ class NewModelBackend(ModelBackend):
     pass
 
 
+class SomeModelBackend(object):
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: Use a more descriptive name, e.g. `BackendWithoutGetUserMethod`.

File: tests/auth_tests/test_auth_backends.py [hunk 4]
```diff
@@ -574,6 +578,20 @@ def test_changed_backend_settings(self):
             self.assertIsNotNone(user)
             self.assertTrue(user.is_anonymous)
 
+    @override_settings(AUTHENTICATION_BACKENDS=[
+        'auth_tests.test_auth_backends.SomeModelBackend',
+        'django.contrib.auth.backends.ModelBackend'])
+    def test_missing_get_user(self):
+        """
+        In force_login() use only authentication backend which has get_user() method.
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: force_login() skips authentication backends without a get_user() method.
(wrap docstrings at 79 characters per the style guide)

File: tests/auth_tests/test_auth_backends.py [hunk 5]
```diff
@@ -574,6 +578,20 @@ def test_changed_backend_settings(self):
             self.assertIsNotNone(user)
             self.assertTrue(user.is_anonymous)
 
+    @override_settings(AUTHENTICATION_BACKENDS=[
+        'auth_tests.test_auth_backends.SomeModelBackend',
+        'django.contrib.auth.backends.ModelBackend'])
+    def test_missing_get_user(self):
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: This test is in the wrong location as you're not changing anything in `contrib.auth`. It should go in `test_client` or test_client_regress`, wherever `force_login()` is tested.

File: tests/auth_tests/test_auth_backends.py [hunk 6]
```diff
@@ -574,6 +578,20 @@ def test_changed_backend_settings(self):
             self.assertIsNotNone(user)
             self.assertTrue(user.is_anonymous)
 
+    @override_settings(AUTHENTICATION_BACKENDS=[
```
Comment (timgraham) [inline tests/auth_tests/test_auth_backends.py]: style:
```
@override_settings(AUTHENTICATION_BACKENDS=[
    'auth_tests.test_auth_backends.SomeModelBackend',
    'django.contrib.auth.backends.ModelBackend',
])
```

File: django/test/client.py (hunk 2 above)
Comment (desecho) [inline django/test/client.py]: No, I just added one test where get_user is completely missing.

**Comment by desecho** [comment]
I've applied changes.

File: django/test/client.py (hunk 2 above)
Comment (timgraham) [inline django/test/client.py]: It seems appropriate to remove `and callable(backend.get_user)` then.

File: django/test/client.py (hunk 1 above)
Comment (timgraham) [inline django/test/client.py]: Yes, I think Django would be obviously broken in such a configuration anyway.

File: tests/test_client/tests.py [hunk 7]
```diff
@@ -549,6 +550,21 @@ def test_force_login_without_backend(self):
...
+        'test_client.auth_backends.BackendWithoutGetUserMethod',
+        'django.contrib.auth.backends.ModelBackend'
+    ])
+    def test_force_login_with_backend_missing_get_user(self):
+        """
+        In force_login() use only authentication backend which has get_user()
+        method.
+        """
+        self.client.force_login(self.u1)
+        request = HttpRequest()
```
Comment (timgraham) [inline tests/test_client/tests.py]: Unless I'm missing the purpose of the rest of the test, it seems sufficient to replace from here down with `self.assertEqual(self.u1.backend, 'django.contrib.auth.backends.ModelBackend')`.

File: tests/test_client/tests.py [hunk 8]
```diff
@@ -549,6 +550,21 @@ def test_force_login_without_backend(self):
         self.assertEqual(response.context['user'].username, 'testclient')
         self.assertEqual(self.u1.backend, 'django.contrib.auth.backends.ModelBackend')
 
+    @override_settings(AUTHENTICATION_BACKENDS=[
+        'test_client.auth_backends.BackendWithoutGetUserMethod',
+        'django.contrib.auth.backends.ModelBackend'
```
Comment (timgraham) [inline tests/test_client/tests.py]: Our convention is to include a trailing comma in places like this so if more items to the list later, we don't have to modify this line again.

File: tests/test_client/tests.py [hunk 9]
```diff
@@ -549,6 +550,21 @@ def test_force_login_without_backend(self):
...
+    ])
+    def test_force_login_with_backend_missing_get_user(self):
+        """
+        In force_login() use only authentication backend which has get_user()
```
Comment (timgraham) [inline tests/test_client/tests.py]: reword: "force_login() skips authentication backends without a get_user() method."

**Pull request merged**